### ISSUE TITLE
[Bug]: Handling of Enums (Storybook 7)

### ISSUE THREAD

**Issue by colangeloproductions** [storybookjs/storybook#23200]
### Describe the bug

I have this Enum for my TextSizes:

```
export enum TextSize {
  small = 'text--size-small',
  medium = 'text--size-medium',
  large = 'text--size-large',
  extralarge = 'text--size-x-large',
}
```

When declaring Enums, its also a Best Practice to use them with a purpose:

```
<Text size={TextSize.large}></Text>
```

But for some reason Storybook doesn't seem to be able to handle the Enum, as Control Values are always set as strings in the "Copy Code" Area:
![image](https://github.com/storybookjs/storybook/assets/93798668/87e87574-d9e3-48b7-ba8b-7b568c31b799)

This is my Text.stories.tsx to show the possible Enum Options:

```
const meta = {
  title: 'Foundations/Text',
  component: Text,
  tags: ['autodocs'],
  argTypes: {
    size: {
      control: 'select',
      options: Object.values(TextSize),
    },
    ...argTypesForIconInText,
  },
} satisfies Meta<typeof Text>;
```

Basically what i would like, or am expecting, is that Storybook is capable of changing the output of the control to be `EnumName.EnumKey`... in my case:

```
<Text size={TextSize.large}></Text>
```

Or am I missing something here that Storybook is already capable of doing...?

**Comment by sookmax** [comment]
I'll reiterate what I commented on [#23212](https://github.com/storybookjs/storybook/issues/23212#issuecomment-1611528254) since this issue seems similar in nature.

To my understanding, Storybook V7 extracts the code snippet, which is `string` from `React.ReactElement`, which is the return value of your component that you are writing stories about (`Text` in your case), when called with the provided args for the story. So by the time the code snippet is extracted, the args of your story (or the props of your component, same thing) will already have been _evaluated_ to values. That's why you're seeing your enum's _value_ (`text--size-medium`) instead of its _expression_ (`TextSize.medium`).

You can workaround this behavior by providing a literal code string to `parameters.docs.source.code`.

For example something like this:

```tsx
...

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  parameters: {
    docs: {
      source: {
        // https://storybook.js.org/docs/react/api/doc-block-source#code
        code: `<Text size={TextSize.large}></Text>`,
        // https://storybook.js.org/docs/react/api/doc-block-source#format
        format: 'dedent',
      },
    },
  },
};
```

**Comment by colangeloproductions** [comment]
@sookmax I see... but that means the one of the most useful storybook feature "copy code" after interacting with the controls, doesnt work, right?

**Comment by sookmax** [comment]
@colangeloproductions Sorry man, I don't think I'm the right person to ask that question. I'm just a community member like you, trying to help myself and others solve/work around the issues they're facing. The general directions or feature sets the core team is planning on bringing is beyond my scope.

**Comment by colangeloproductions** [comment]
@sookmax all good man! appreciate the help!

**Comment by meriturva** [comment]
Hi to all,
any news here, it isn't very pleasant to define `argTypes` for every enum (storybook v8.0).

Here is a sample I would like to handle automatically:

```export default {
  component: MyButtonComponent,
  subcomponents: { MyComponentDirective },
  tags: ['autodocs'],
  title: 'My/Button',
  argTypes: {
    size: {
      control: 'radio',
      options: Object.values(MyButtonSize)
    }
  }
} as Meta<MyButtonComponent>;```

**Comment by creator-woow** [comment]
> Hi to all, any news here, it isn't very pleasant to define `argTypes` for every enum (storybook v8.0).
> 
> Here is a sample I would like to handle automatically:
> 
> ```
>   component: MyButtonComponent,
>   subcomponents: { MyComponentDirective },
>   tags: ['autodocs'],
>   title: 'My/Button',
>   argTypes: {
>     size: {
>       control: 'radio',
>       options: Object.values(MyButtonSize)
>     }
>   }
> } as Meta<MyButtonComponent>;```
> ```


Hi, have you found a way to avoid manually specifying the argument type?

**Comment by meriturva** [comment]
> > Hi to all, any news here, it isn't very pleasant to define `argTypes` for every enum (storybook v8.0).
> > Here is a sample I would like to handle automatically:
> > ```
> >   component: MyButtonComponent,
> >   subcomponents: { MyComponentDirective },
> >   tags: ['autodocs'],
> >   title: 'My/Button',
> >   argTypes: {
> >     size: {
> >       control: 'radio',
> >       options: Object.values(MyButtonSize)
> >     }
> >   }
> > } as Meta<MyButtonComponent>;```
> > ```
> 
> [@meriturva](https://github.com/meriturva) Hi, have you found a way to avoid manually specifying the argument type?

@creator-woow 
Not yet!

**Comment by creator-woow** [comment]
@meriturva Found it for you. Need to change your storybook's main.ts file like this![Image](https://github.com/user-attachments/assets/e56ffec3-f75f-4731-bae1-2bd707cdfc16)

**Comment by NicoleLatifi** [comment]
> [@sookmax](https://github.com/sookmax) I see... but that means the one of the most useful storybook feature "copy code" after interacting with the controls, doesnt work, right?

I was facing the same issue and found a solution using the `transform` property (instead of `code`) and `source.replace()`. I.e. you could do something like:

```
export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  parameters: {
    docs: {
      source: {
        // https://storybook.js.org/docs/react/api/doc-block-source#transform
        transform: (source: string) => {
          return source.replace(
                    /textSize="([^"]+)"/g,
                    (_, match) => {
                        // Find the matching key in TextSize
                        const textSizeKey = Object.keys(TextSize).find(
                            key => TextSize[key] === match
                        );
                        return textSizeKey ? `textSize={TextSize.${textSizeKey}}` : `textSize="${match}"`;
                    }
                );
        }
      },
    },
  },
};
```
